// @sap-ux/store/dist/i18n.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initI18n = initI18n;
exports.text = text;
const i18next_1 = __importDefault(require("i18next"));
const ux_store_i18n_json_1 = __importDefault(require("./translations/ux-store.i18n.json"));
const NS = 'ux-store';
let i18nInstance = i18next_1.default.createInstance();
async function initI18n() {
    i18nInstance = i18next_1.default.createInstance({
        resources: {
            en: {
                [NS]: ux_store_i18n_json_1.default
            }
        },
        lng: 'en',
        fallbackLng: 'en',
        defaultNS: NS,
        fallbackNS: 'default',
        ns: [NS],
        interpolation: { escapeValue: false }
    });
    await i18nInstance.init();
}
function text(key, options) {
    return i18nInstance.t(key, typeof options === 'string' ? { defaultValue: options } : options);
}
