// @sap-ux/fiori-generator-shared/dist/types/app-gen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FloorplanFE = exports.FloorplanFF = void 0;
// Union types to expose a single interface property for Floorplan
// This provides a layer of abstraction to isolate internal changes from external headless API consumers
// Since these keys are used as an external API definiton they need to be meaningful
// Note that ordering here determines rendering order
/**
 * Due to ts(18033) we cannot use the type values directly here:
 * FF_SIMPLE = FFTemplateType.Basic // Once https://github.com/microsoft/TypeScript/pull/59475 is merged we can remove the hardcoded values and directly use the template values
 */
var FloorplanFF;
(function (FloorplanFF) {
    FloorplanFF["FF_SIMPLE"] = "basic";
})(FloorplanFF || (exports.FloorplanFF = FloorplanFF = {}));
/**
 * Due to ts(18033) we cannot use the type values directly here:
 * Once https://github.com/microsoft/TypeScript/pull/59475 is merged we can remove hardcoded values and directly use the template values
 * FE_FPM = FETemplateType.FlexibleProgrammingModel,
 * FE_LROP = FETemplateType.ListReportObjectPage,
 * FE_OVP = FETemplateType.OverviewPage,
 * FE_ALP = FETemplateType.AnalyticalListPage,
 * FE_FEOP = FETemplateType.FormEntryObjectPage,
 * FE_WORKLIST = FETemplateType.Worklist
 */
var FloorplanFE;
(function (FloorplanFE) {
    FloorplanFE["FE_FPM"] = "fpm";
    FloorplanFE["FE_LROP"] = "lrop";
    FloorplanFE["FE_OVP"] = "ovp";
    FloorplanFE["FE_ALP"] = "alp";
    FloorplanFE["FE_FEOP"] = "feop";
    FloorplanFE["FE_WORKLIST"] = "worklist";
})(FloorplanFE || (exports.FloorplanFE = FloorplanFE = {}));
