// @sap-ux/odata-service-writer/dist/i18n.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.i18n = void 0;
exports.initI18n = initI18n;
exports.t = t;
const i18next_1 = __importDefault(require("i18next"));
const odata_service_writer_i18n_json_1 = __importDefault(require("./translations/odata-service-writer.i18n.json"));
const NS = 'odata-service-writer';
exports.i18n = i18next_1.default.createInstance();
/**
 * Initialize i18next with the translations for this module.
 */
async function initI18n() {
    await exports.i18n.init({
        resources: {
            en: {
                [NS]: odata_service_writer_i18n_json_1.default
            }
        },
        lng: 'en',
        fallbackLng: 'en',
        defaultNS: NS,
        ns: [NS]
    });
}
/**
 * Helper function facading the call call to i18next.
 *
 * @param key i18n key
 * @param options additional options
 * @returns {string} localized string stored for the given key
 */
function t(key, options) {
    return exports.i18n.t(key, options);
}
initI18n().catch(() => {
    // Ignore any errors since the write will still work
});
