// @sap-ux/axios-extension/dist/abap/catalog/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CatalogService = exports.ServiceType = void 0;
const odata_service_1 = require("../../base/odata-service");
exports.ServiceType = {
    UI: 'UI',
    NotClassified: 'Not Classified',
    WebApi: 'WEB_API',
    NotDetermined: 'Not Determined'
};
/**
 * OData version independent abstract base class of SAP's catalog service
 */
class CatalogService extends odata_service_1.ODataService {
    entitySet;
    services;
    isS4Cloud;
    /**
     * Returns list of services from the catalog service.
     *
     * @param useNextLink if true, the next link tags will be used to fetch the next page of results, pages are fetched serially.
     * Note that this will be less performant for larger datasets.
     * @returns list of services
     */
    async listServices(useNextLink = false) {
        if (!this.services) {
            this.services = await this.fetchServices(useNextLink);
        }
        return this.services;
    }
}
exports.CatalogService = CatalogService;
